lib/context.js

```diff
@@ -46,37 +66,41 @@ class ResponseContext {
     return urlJoin(this._config.baseURL, this._config.redirectUriPath);
   }
 
-  async login(params = {}) {
+  async login(options = {}) {
     const next = cb(this._next).once();
     const req = this._req;
     const res = this._res;
     const config = this._config;
-
     const client = req.openid.client;
-    const authorizeParams = config.authorizationParams;
+
+    options = {
+      returnTo: this._config.baseURL,
+      authorizationParams: {},
+      ...options
+    };
+
+    options.authorizationParams = {
+      redirect_uri: this.getRedirectUri(),
+      ...config.authorizationParams,
+      ...options.authorizationParams
+    };
+
     const transientOpts = {
       legacySameSiteCookie: config.legacySameSiteCookie,
-      sameSite: config.authorizationParams.response_mode === 'form_post' ? 'None' : 'Lax'
+      sameSite: options.authorizationParams.response_mode === 'form_post' ? 'None' : 'Lax'
+    };
+
+    const stateTransientOpts = {
+      ...transientOpts,
+      value: config.getLoginState(req, options)
     };
 
     try {
-      let returnTo;
-      if (params.returnTo) {
-        returnTo = params.returnTo;
-      } else if (req.method === 'GET') {
-        returnTo = req.originalUrl;
-      } else {
-        returnTo = this._config.baseURL;
-      }
-
-      // TODO: Store this in state
-      transient.store('returnTo', res, Object.assign({value: returnTo}, transientOpts));
-
-      const authParams = Object.assign({
+      const authParams = {
+        ...options.authorizationParams,
```

Comment:
Subtle change here is that `nonce` and `state` passed in using the `options` object are ignored. The `state` parameter is the only one that really needs the ability to be developer-defined, which is the whole purpose of this PR. Giving another, unnecessary way to do custom state would be confusing and I don't see a lot of value in allowing a custom `nonce`. Both would require the developer to store it in cookies in a way the library can read on the callback.